source/extensions/filters/network/zookeeper_proxy/decoder.h

```diff
@@ -154,8 +156,11 @@ class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
   void decodeAndBufferHelper(Buffer::Instance& data, DecodeType dtype,
                              Buffer::OwnedImpl& zk_filter_buffer);
   void decode(Buffer::Instance& data, DecodeType dtype, uint64_t full_packets_len);
-  void decodeOnData(Buffer::Instance& data, uint64_t& offset);
-  void decodeOnWrite(Buffer::Instance& data, uint64_t& offset);
+  // The return value of decodeOnData and decodeOnWrite will be ZooKeeper opcode or absl::nullopt.
+  // absl::nullopt indicates WATCH_XID, which is generated by the server and has no corresponding
```

Comment:
did we test WATCH_XID?

Reply:
Yeah, we tested here: https://github.com/envoyproxy/envoy/pull/29081/files#diff-a0f6243d3325c5bb6bbda5bb0e9c6a47723f25d95934e2edf05a98dc978f8f4fR1373-R1390